agent/agent_endpoint.go

```diff
@@ -380,7 +380,10 @@ func (s *HTTPHandlers) AgentServices(resp http.ResponseWriter, req *http.Request
 // blocking watch using hash-based blocking.
 func (s *HTTPHandlers) AgentService(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
 	// Get the proxy ID. Note that this is the ID of a proxy's service instance.
-	id := strings.TrimPrefix(req.URL.Path, "/v1/agent/service/")
+	id, err := getPathSuffixUnescaped(req.URL.Path, "/v1/agent/service/")
+	if err != nil {
+		s.agent.logger.Error(err.Error())
```

Comment:
When I ran against a consul binary with an invalid encoded value, this is what happened:
```
$ curl --request GET localhost:8500/v1/agent/service/counting%2id
400 Bad Request

# No error appears in consul log output
```

@kisunji : Is this what we'd expect/want to happen?

Reply:
Summarizing our discussion over slack: most middlewares reject invalid URLs early so the request never reaches the handler. It's almost impossible to externally test the error check via an HTTP client because you either: 1) face early rejection by go's http lib; or 2) encode into a valid url and never hit the error code path.